fe/fe-core/src/main/java/com/starrocks/meta/lock/LockManager.java

```diff
@@ -0,0 +1,521 @@
+                        break;
+                    }
+
+                    if (Config.dead_lock_detection) {
+                        victim = checkAndHandleDeadLock(rid, locker, lockType);
+                    }
+                    if (victim != null) {
+                        /* Locker owns the lock and no deadlock was detected. */
+                        break;
+                    }
+
+                    try {
+                        if (timeout == 0) {
+                            locker.wait(0);
+                        } else {
+                            locker.wait(Math.max(1, timeRemain(timeout, startTime)));
+                        }
+                    } catch (InterruptedException ie) {
+                        removeFromWaiterList(rid, locker, lockType);
+                        throw new LockInterruptException(ie);
+                    }
+
+                    //locker is wakeup normally and becomes the owner
+                    if (isOwner(rid, locker, lockType)) {
+                        break;
+                    }
+
+                    boolean lockTimeOut = (timeout != 0) && timeRemain(timeout, startTime) < 0;
+                    if (lockTimeOut) {
+                        removeFromWaiterList(rid, locker, lockType);
+
+                        /* Failure to acquire lock within the timeout ms*/
+                        throw new LockTimeoutException("");
+                    }
+
+                    /*
+                     * There are two reasons for the loop below.
+                     *
+                     * 1. When another thread detects a deadlock and notifies this thread,
+                     * it will wakeup before the timeout interval has expired. We must loop
+                     * again to perform deadlock detection. Normally, if the deadlock
+                     * detected by the other thread is still present, this locker will be
+                     * selected as the victim and we will throw DeadLockException below.
+                     *
+                     * 2. spurious wakeups
+                     */
+                }
+
+                if (victim == null) {
+                    assert isOwner(rid, locker, lockType);
+                    locker.clearWaitingFor();
+                }
+            }
+
+            if (victim == null) {
+                /* Locker owns the lock and no deadlock was detected. */
+                return;
+            } else {
+                /*
+                 * A deadlock is detected and this locker is not the victim.
+                 * Notify the victim.
+                 */
+                boolean currentLockerIsOwner = notifyVictim(victim, locker, rid, lockType, timeout, startTime);
+                if (currentLockerIsOwner) {
+                    synchronized (locker) {
+                        locker.clearWaitingFor();
+                    }
+                    return;
+                }
+
+                /* Retry */
```

Comment:
plz add more comments, this means that after notify the victim, current locker still cannot get the lock and need to wait to be notified again, but in this case we should not do dead lock detection immediately.